### PR TITLE
feat: new custom auth/link handler using argus id

### PR DESCRIPTION
**Pull request by rmrt1n**
Closes: WORLD-1254

## Overview

This PR adds the handlers for custom authentication and linking using [Argus ID](https://app.campsite.com/argus-labs/notes/r3v47d5ui0vw). ATM, only us (Argus Labs) can use this custom authentication as it requires a symmetric JWT secret key to verify the JWTs issued by Supabase. Once [asymmetric JWTs](https://github.com/orgs/supabase/discussions/29289) are rolled out, third parties should be able to use it too.

There's a slight change from the original design document. Instead of using the JWTs as the IDs for `AuthenticateCustom`, we're using the SHA256 hash of the JWT as the ID and including the JWT in the request variables. This change is made because `AuthenticateCustom` has an ID max length limit of 128 characters, which isn't long enough to contain the entirety of the JWT.

## Brief Changelog

- Add new custom auth type `'argus'` and the handlers for authentication and linking.
- Add logic for verifying and parsing the JWT.
- Add trace information to both Argus ID and SIWE handlers.

## Testing and Verifying

Added unit tests to validate the JWT verifying and parsing logic.

## Release Notes

- **New Features**
	- Introduced Argus ID authentication using JWT for enhanced user authentication.
	- Added functionality for linking accounts with Argus ID.

- **Improvements**
	- Enhanced error handling and logging for custom authentication processes.
	- Added OpenTelemetry tracing to authentication and linking functions for better observability.
	- Updated SIWE validation error messages for improved clarity.
	- Implemented a check for JWT secret during initialization to prevent errors.

- **Tests**
	- Implemented comprehensive unit tests for JWT validation and parsing to ensure reliability.

### CODE REVIEW

**Comment by rmrt1n** [comment]
* **#815** <a href="https://app.graphite.dev/github/pr/Argus-Labs/world-engine/815?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Argus-Labs/world-engine/815?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: relay/nakama/auth/argus_id.go
Comment (rmrt1n) [reply relay/nakama/auth/argus_id.go:112]: this is fine

File: relay/nakama/auth/custom.go
Comment (rmrt1n) [reply relay/nakama/auth/custom.go:44]: it's fine

File: relay/nakama/auth/custom.go
Comment (rmrt1n) [reply relay/nakama/auth/custom.go:55]: this should be covered by errors returned in `validateAndParseJWT`

File: relay/nakama/auth/argus_id.go
Comment (rmrt1n) [reply relay/nakama/auth/argus_id.go:165]: I'm pretty sure its set to empty map by default

File: relay/nakama/auth/argus_id_test.go
Comment (rmrt1n) [reply relay/nakama/auth/argus_id_test.go:102]: it's fine

File: relay/nakama/auth/custom.go
Comment (rmrt1n) [reply relay/nakama/auth/custom.go:36]: it's not called concurrently

**Review by smsunarto — CHANGES_REQUESTED** [review]
just some nits, otherwise lgtm

File: relay/nakama/auth/argus_id.go
Comment (rmrt1n) [reply relay/nakama/auth/argus_id.go:31]: Found out that `claims, ok := token.Claims.(*SupabaseClaims)` isn't a strict type coersion, and will return ok for claims that look nothing like the desired type. This means we still have to manually check the existence of every field that we need.

I ended up removing all of fields that are not used, which leaves:

```go
type SupabaseClaims struct {
  jwt.StandardClaims
  UserMetaData map[string]interface{} `json:"user_metadata"`
}
```
wdyt?

Comment (smsunarto) [reply relay/nakama/auth/argus_id.go:31]: lgtm

File: relay/nakama/auth/argus_id.go
Comment (smsunarto) [reply relay/nakama/auth/argus_id.go]: @rmrt1n I don't think this is resolved yet.

File: relay/nakama/auth/siwe.go
Comment (smsunarto) [inline relay/nakama/auth/siwe.go:89]: you don't need to be dependency injecting the span around, you can use SpanFromContext (see https://opentelemetry.io/docs/languages/go/instrumentation/#get-the-current-span)

**Review by smsunarto — CHANGES_REQUESTED** [review]

File: relay/nakama/auth/argus_id.go
Comment (rmrt1n) [reply relay/nakama/auth/argus_id.go]: I did though, [here](https://github.com/Argus-Labs/world-engine/pull/815/files#diff-9d382e42476b01f2cc3dfa8e13d43136df75cfa689c56f6a4ca30d32ef6dd57eR44).

File: relay/nakama/auth/custom.go
Comment (smsunarto) [inline relay/nakama/auth/custom.go:80]: This should also be a subspan instead of event

File: relay/nakama/auth/custom.go
Comment (smsunarto) [inline relay/nakama/auth/custom.go:69]: This should also be a subspan instead of event

**Review by smsunarto — APPROVED** [review]
Approving with comments -- there are some that is still using OTel trace event that should've been subspan. Once that's fixed feel free to merge

**Pull request merged**
